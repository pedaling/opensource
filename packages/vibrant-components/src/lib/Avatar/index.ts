export { Avatar } from './Avatar';
export type { AvatarProps } from './AvatarProps';
