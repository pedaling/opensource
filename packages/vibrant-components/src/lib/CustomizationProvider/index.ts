export type { CustomizationProviderProps, Configurations } from './CustomizationProvider';
export { CustomizationProvider, useCustomization } from './CustomizationProvider';
