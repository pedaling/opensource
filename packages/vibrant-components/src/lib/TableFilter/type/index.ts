export type StringFilterOperator = 'contains' | 'empty' | 'equals' | 'notContains' | 'notEmpty' | 'notEquals';

export type StringFilter = {
  value: string;
  operator: StringFilterOperator[];
};

export type Filter = StringFilter & { dataKey: string };
