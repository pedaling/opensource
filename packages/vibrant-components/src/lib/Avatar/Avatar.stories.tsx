import type { ComponentMeta, ComponentStory } from '@storybook/react';
import { Avatar } from './Avatar';

export default {
  title: 'Avatar',
  component: Avatar,
  args: {
    size: 'lg',
    src: 'https://cdn.class101.net/images/7776e542-d6db-49c6-b86f-7e93f0cf5425/200x200.png',
    placeholder: 'https://cdn.class101.net/images/2483a7e5-8c20-4ac7-a209-805b4482664d',
    alt: 'avatar',
  },
} as ComponentMeta<typeof Avatar>;

export const Basic: ComponentStory<typeof Avatar> = props => <Avatar {...props} />;

export const Placeholder: ComponentStory<typeof Avatar> = props => <Avatar {...props} src="" />;
