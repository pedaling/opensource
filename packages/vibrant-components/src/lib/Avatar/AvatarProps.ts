import type { ResponsiveValue } from '@vibrant-ui/core';
import { propVariant, withVariation } from '@vibrant-ui/core';

export type AvatarProps = {
  testId?: string;
  size: ResponsiveValue<number | 'lg' | 'md' | 'sm' | 'xs'>;
  src: string;
  alt: string;
  placeholder?: string;
};

export const withAvatarVariation = withVariation<AvatarProps>('Avatar')(
  propVariant({
    props: [
      {
        name: 'size',
        responsive: true,
      },
    ],
    variants: ({ size }) => {
      if (typeof size === 'number') {
        return {
          size,
          borderRadius: size / 2,
        };
      }

      if (size === 'xs') {
        return { size: 16, borderRadius: 8 };
      }

      if (size === 'sm') {
        return { size: 24, borderRadius: 12 };
      }

      if (size === 'md') {
        return { size: 30, borderRadius: 15 };
      }

      return { size: 44, borderRadius: 22 };
    },
  })
);
