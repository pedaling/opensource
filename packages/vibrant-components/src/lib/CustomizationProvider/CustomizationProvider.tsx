import type { ComponentType, FC } from 'react';
import { createContext, useContext, useMemo } from 'react';
import type { ReactElementChild, ResponsiveValue } from '@vibrant-ui/core';

export type Configurations = {
  avatar?: {
    placeholder: string;
  };
  cardNumberField?: {
    cardIconMap: Record<string, ComponentType<{ size: ResponsiveValue<number> }>>;
  };
};

type CustomizationContextValue = Configurations;

const CustomizationContext = createContext<CustomizationContextValue>({
  avatar: {
    placeholder: 'https://cdn.class101.net/images/e1cba897-73d1-43de-864b-c36cefdea670/200x200.png',
  },
  cardNumberField: {
    cardIconMap: {},
  },
});

export type CustomizationProviderProps = Configurations & { children: ReactElementChild };

export const CustomizationProvider: FC<CustomizationProviderProps> = ({ children, ...props }) => {
  const config = useCustomization();

  const mergedCustomization = useMemo<CustomizationContextValue>(
    () => ({
      ...config,
      ...props,
    }),
    [config, props]
  );

  return <CustomizationContext.Provider value={mergedCustomization}>{children}</CustomizationContext.Provider>;
};

export const useCustomization = (): CustomizationContextValue => useContext(CustomizationContext);

CustomizationProvider.displayName = 'CustomizationProvider';
