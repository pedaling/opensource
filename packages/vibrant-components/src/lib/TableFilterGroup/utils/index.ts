import type { DateFilterOperator, MultiSelectFilterOperator, Option, StringFilterOperator } from '../types';

export const isValueRequiredOperator = (op: string) => op === 'empty' || op === 'notEmpty';

export const isStringFilterValid = (filter: { value: string; operator: StringFilterOperator }) =>
  Boolean(filter.value) || isValueRequiredOperator(filter.operator);

export const isDateFilterValid = (filter: { value: Date[]; operator: DateFilterOperator }) => {
  if (isValueRequiredOperator(filter.operator)) return true;

  if (filter.operator === 'between') return filter.value.length >= 2;

  return filter.value.length >= 1;
};

export const isMultiSelectFilterValid = (filter: { value: Option['value'][]; operator: MultiSelectFilterOperator }) =>
  Boolean(filter.value.length > 0) || isValueRequiredOperator(filter.operator);
