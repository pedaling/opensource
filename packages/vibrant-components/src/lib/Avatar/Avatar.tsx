import { useState } from 'react';
import { Image } from '@vibrant-ui/core';
import { useCustomization } from '../CustomizationProvider';
import { withAvatarVariation } from './AvatarProps';

export const Avatar = withAvatarVariation(({ testId, src, alt, size, placeholder }) => {
  const { avatar } = useCustomization();
  const [fallbackSrc, setFallbackSrc] = useState<string>();

  const handleError = () => {
    setFallbackSrc(placeholder ?? avatar?.placeholder);
  };

  return (
    <Image
      testId={testId}
      src={fallbackSrc ?? src}
      alt={alt}
      width={size}
      height={size}
      borderRadius="50%"
      onError={handleError}
    />
  );
});
