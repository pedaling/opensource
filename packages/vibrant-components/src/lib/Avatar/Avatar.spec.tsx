import { waitFor } from '@testing-library/react';
import type { ReactRenderer } from '@vibrant-ui/utils/testing-web';
import { createReactRenderer } from '@vibrant-ui/utils/testing-web';
import { CustomizationProvider } from '../CustomizationProvider';
import { Avatar } from './Avatar';

describe('<Avatar />', () => {
  const { render } = createReactRenderer();
  let renderer: ReactRenderer;

  describe('when size is lg', () => {
    beforeEach(async () => {
      renderer = render(<Avatar size="lg" src="" alt="" />);
    });

    it('match snapshot', () => {
      expect(renderer.container).toMatchSnapshot();
    });
  });

  describe('when size is md', () => {
    beforeEach(async () => {
      renderer = render(<Avatar size="md" src="" alt="" />);
    });

    it('match snapshot', () => {
      expect(renderer.container).toMatchSnapshot();
    });
  });

  describe('when size is sm', () => {
    beforeEach(async () => {
      renderer = render(<Avatar size="sm" src="" alt="" />);
    });

    it('match snapshot', () => {
      expect(renderer.container).toMatchSnapshot();
    });
  });

  describe('when size is xs', () => {
    beforeEach(async () => {
      renderer = render(<Avatar size="xs" src="" alt="" />);
    });

    it('match snapshot', () => {
      expect(renderer.container).toMatchSnapshot();
    });
  });

  describe('when valid src provided', () => {
    beforeEach(() => {
      renderer = render(
        <Avatar
          size="md"
          src="https://cdn.class101.net/images/7776e542-d6db-49c6-b86f-7e93f0cf5425/200x200.png"
          alt=""
        />
      );
    });

    it('should render the img element with src as source', () => {
      expect(renderer.getByRole('img').getAttribute('src')).toBe(
        'https://cdn.class101.net/images/7776e542-d6db-49c6-b86f-7e93f0cf5425/200x200.png'
      );
    });
  });

  describe('when invalid src provided', () => {
    const defaultPlaceholder = 'https://cdn.class101.net/images/e1cba897-73d1-43de-864b-c36cefdea670/200x200.png';

    beforeEach(() => {
      renderer = render(
        <CustomizationProvider avatar={{ placeholder: defaultPlaceholder }}>
          <Avatar size="md" src="" alt="" />
        </CustomizationProvider>
      );
    });

    it('should render the img element with default placeholder as source', async () => {
      await waitFor(() => expect(renderer.getByRole('img').getAttribute('src')).toBe(defaultPlaceholder));
    });
  });

  describe('when alt provided', () => {
    beforeEach(() => {
      renderer = render(
        <Avatar
          size="md"
          src="https://cdn.class101.net/images/7776e542-d6db-49c6-b86f-7e93f0cf5425/200x200.png"
          alt="avatar"
        />
      );
    });

    it('should render the img element with alt', () => {
      expect(renderer.queryByAltText('avatar')).toBeTruthy();
    });
  });

  describe('when placeholder provided and invalid src provided', () => {
    const placeholder = 'https://cdn.class101.net/images/e1cba897-73d1-43de-864b-c36cefdea670/100x100.png';

    beforeEach(() => {
      renderer = render(<Avatar size="md" src="" alt="" placeholder={placeholder} />);
    });

    it('should render the img element with placeholder value as source', async () => {
      await waitFor(() => expect(renderer.getByRole('img').getAttribute('src')).toBe(placeholder));
    });
  });
});
