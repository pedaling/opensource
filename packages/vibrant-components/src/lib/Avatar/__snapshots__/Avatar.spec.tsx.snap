// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Avatar /> when size is lg match snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
  width: 44px;
  height: 44px;
  border-radius: 22px;
}

<div>
  <img
    alt=""
    class="emotion-0"
    data-testid="image"
    draggable="false"
    loading="eager"
    src="https://cdn.class101.net/images/2483a7e5-8c20-4ac7-a209-805b4482664d"
  />
</div>
`;

exports[`<Avatar /> when size is md match snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
  width: 30px;
  height: 30px;
  border-radius: 15px;
}

<div>
  <img
    alt=""
    class="emotion-0"
    data-testid="image"
    draggable="false"
    loading="eager"
    src="https://cdn.class101.net/images/2483a7e5-8c20-4ac7-a209-805b4482664d"
  />
</div>
`;

exports[`<Avatar /> when size is sm match snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
  width: 24px;
  height: 24px;
  border-radius: 12px;
}

<div>
  <img
    alt=""
    class="emotion-0"
    data-testid="image"
    draggable="false"
    loading="eager"
    src="https://cdn.class101.net/images/2483a7e5-8c20-4ac7-a209-805b4482664d"
  />
</div>
`;

exports[`<Avatar /> when size is xs match snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-sizing: border-box;
  position: relative;
  width: 16px;
  height: 16px;
  border-radius: 8px;
}

<div>
  <img
    alt=""
    class="emotion-0"
    data-testid="image"
    draggable="false"
    loading="eager"
    src="https://cdn.class101.net/images/2483a7e5-8c20-4ac7-a209-805b4482664d"
  />
</div>
`;
